### PR TITLE
Add Dataset Curation Tools (FPS & K-Means) for LoRA Training

### PR DESCRIPTION
**Pull request by NMWave**
**Description:**
This PR adds a "Curator" module to PhotoMap, designed to filter large image datasets down to high-quality subsets for AI model training (LoRA/Fine-tuning).

**New Features:**
**Diversity Sampling (FPS):** Uses Farthest Point Sampling on CLIP embeddings to select visually distinct images (outliers/edges).
**Representative Sampling (K-Means):** Clusters images and picks centroids to find the most representative examples of the dataset.
**Smart Export:** Exports the selected subset (images + text files) to a target folder, handling filename collisions and subfolders automatically.
**Visual Feedback:** Highlights selected images in the Grid View and draws specific markers on the UMAP visualization.

**Technical Changes:**
Added backend/routers/curation.py.
Refactored caching in embeddings.py to allow global function access (needed for stateless API calls).
Added Curation Panel to main.html.

I made this to use on my own datasets to help curate them for better diversity for lora training (to ensure prevent overcooking a concept) and thought I'd share as it'll be helpful for others. Here's an example in action, 500+ images, and I can export the 30 most diverse from this set for use in finetuning/lora training.

image

### CODE REVIEW

**Comment by NMWave** [comment]
> This is very cool and I'll review your contribution within the next couple of days. I'm curious about how you found PhotoMap? I've done almost nothing to promote it.

Thank you! I was hunting for a data science tool to help curate image datasets for model training and stumbled into your post on reddit.

My original plan was to build my own tool from scratch, but seeing as how you already had all the visual elements and clip embedding down, I just needed to build the data science stuff on top, so it was easier to do it this way (I'm u/AcadiaVivid on reddit).

image

Refer to the image. With this, the user will be able to:
1) Run a monte carlo simulation with their preferred algorithm for x iterations, to select the top n images.
2) Once the most diverse images from their dataset, then use manual refinement (>90% for instance) to lock outliers, and run the simulation again with those excluded.
3) The user can then export that dataset out (image + text exported together) into the folder of their choice.
4) The user can also get data on the monte carlo  analysis exported into a csv (how many times did each image get selected in the simulation

The selections are color coded to show how many times they were selected and what was locked.

Working perfectly for my use-case, hopefully others in the community can use it too

Will publish my changes sometime later today

**Comment by lstein** [comment]
I'll do a full review over the next couple of days, but I took a quick look and it looks pretty good! In the meantime I asked copilot to do a review and it found a number of small problems (mostly missing docstrings) and a few bugs such as missing event handlers. Could you take a look and address the major issues?

Also I would be obliged if you could update the documentation to describe the curator mode. I think curator mode deserves a new documentation page of its own, likely under docs/user-guide.

Thank you so much for your contribution. I am very excited about merging this into the main branch. And just in time for the 1.0 release too!

**Comment by lstein** [comment]
> Hey added the doc, a test for curation.py router and multiple other fixes/improvements such as to docstrings, hope that helps with the merging process

Fantastic. Going over everything now.

File: photomap/frontend/static/javascript/curation.js [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+        });
+
+        if (filesToExport.length === 0) {
+            alert("No files selected to export (all excluded?).");
+            return;
+        }
+
+        setStatus(`Exporting ${filesToExport.length} files...`, "loading");
+        try {
+            const response = await fetch('/api/curation/export', {
```
Comment (lstein) [inline photomap/frontend/static/javascript/curation.js]:
```suggestion
            const response = await fetch('api/curation/export', {
```
Another place where a relative API URL is needed.

**Comment by lstein** [comment]
> > > Hey added the doc, a test for curation.py router and multiple other fixes/improvements such as to docstrings, hope that helps with the merging process
> > 
> > 
> > Fantastic. Going over everything now.
> 
> Just some thoughts I had last night which could help polish and unify some aspects of the app, feel free to ignore these though:
> 
> * Something small I think would be nice to fix with the base tool: When adding a new album, double clicking to add a new folder is not a fluid experience (it often highlights the text, or brings up a submenu). Maybe make it single click to go into a folder (similar to how going back out of a folder is right now) for a smoother experience.
> * I noticed that when doing a text search (base feature), it highlights the similar images, and dims the non-similar. Can I recommend you use a different color for the similar images (perhaps yellow dots?). When I had 5000+ images loaded up, brightness alone doesn't cut it.
> * It would be good if albums can be toggled on and off in the content manager. Would probably require re-indexing to remap the relationships each time on the umap I assume? Still a useful feature though. While using the curator tool to create a dataset, I find myself deleting and adding different folders often.
> * The curator tool actually exports the text files from the source folder, it would be nice if text files (or captions) had native support, perhaps in the metadata explorer. That way a user can view the captions of an image if it exists (not just metadata), and potentially edit it. You could even add an autogenerate captions with JoyTag feature at a later point in time to make it even better.
> * With Z-Image becoming the go-to lightweight model, once the base model comes out, to aid with training it, it could be beneficial to use the abliterated Qwen3 4B model for embeddings: [https://huggingface.co/huihui-ai/Huihui-Qwen3-VL-4B-Instruct-abliterated](url). That way the relationship map built would be the same as the underlying z-image models understanding of concepts.
> * Feel free to change the UI or button placement for the curator tool to best match the look and feel of PhotoMap

Great suggestions. I particularly like the idea of using the Qwen3 model for the embeddings. I am not 100% happy with the accuracy of the CLIP-based clusterings.

**Comment by lstein** [comment]
I've started my review. I've tested the curator functionality and it appears to work as advertised. However, I still have more review to do.

Here are a few requests/suggestions so far:
1. Could I ask that you rebase against the latest master branch?  I'm seeing lots of code deletions that are actually code that I added in the last set of commits to master, and this is causing at least one regression.
2. I like the behavior of exiting umap full-screen mode when a cluster is clicked. Could you condition the suppression of this behavior against curator mode?
3. You could make the curator dialog box draggable so that the user can reposition it. Suitable dialog-dragging code can be found in metadata-drawer.js and in umap.js itself.
4. I would like to fix the curator mode button to be more consistent with the icon style, but probably best if I tinker with the appearance and placement myself. Could you "Allow edits by maintainer" if it is not already set?

**Comment by lstein** [comment]
> Something small I think would be nice to fix with the base tool: When adding a new album, double clicking to add a new folder is not a fluid experience (it often highlights the text, or brings up a submenu). Maybe make it single click to go into a folder (similar to how going back out of a folder is right now) for a smoother experience.
I don't think I can descend into the folder with a single click because how do I distinguish between the user single-clicking to select the folder as the image location vs single-clicking to descend into the folder to find a more deeply nested one?

I'll see what I can do to make the experience smoother for the user. I don't experience the glitches you describe on my Linux/Chrome system - what browser and OS are you seeing the problems on?

**Comment by NMWave** [comment]
Hey mate apologies, won't be able to commit any time to this over the next few weeks (work + personal commitments getting very close to the end of the year). 

Perhaps merge what's there at the moment as it's hopefully in good shape, apologies if I overwrote some of the functionality from your last commit, this was a very extensive addition that started out for personal use and I thought it'd be worth to share. Noted about the additions, I'll see if I can help over the xmas break, but feel free to implement as well as you see fit.

Allow edits by maintainer is toggled as well.

Regarding the text highlighting glitch, it's occurring on Win11 + Edge

**Comment by lstein** [comment]
Hi! I hope you're having a nice holiday break. Just to let you know that I merged in the curation panel and pushed it out into Photomap version 1.0.0 just now. Thank you so much for the contribution and I hope we can work together in the future.
